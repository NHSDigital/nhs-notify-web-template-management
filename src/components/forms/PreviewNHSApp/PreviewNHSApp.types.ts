export type PreviewNHSAppProps = {
  templateName: string;
  message: string;
};
