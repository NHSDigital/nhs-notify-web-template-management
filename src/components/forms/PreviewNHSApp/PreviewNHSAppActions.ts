import { MarkdownItWrapper } from '@utils/markdownit';
import { Dependencies, IPreviewNHSAppActions } from './PreviewNHSApp.types';

export class PreviewNHSAppActions implements IPreviewNHSAppActions {
  private readonly _markdownClient: MarkdownItWrapper;

  constructor(
    _container: Dependencies = { markdownClient: new MarkdownItWrapper() }
  ) {
    this._markdownClient = _container.markdownClient
      .enableLineBreak()
      .enable(['heading', 'link', 'list', 'emphasis']);
  }

  renderMarkdown(value: string) {
    'use server';
    return this._markdownClient.render(value);
  }
}
