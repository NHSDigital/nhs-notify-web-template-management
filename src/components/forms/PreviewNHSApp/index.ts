export * from './PreviewNHSApp';
export * from './PreviewNHSApp.types';
export * from './PreviewNHSAppActions';
