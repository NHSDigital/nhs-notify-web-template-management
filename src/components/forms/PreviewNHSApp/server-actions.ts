import { MarkdownItWrapper } from '@/src/utils/markdownit';

export function renderMarkdown(
  value: string,
  markdown = new MarkdownItWrapper()
) {
  markdown.enableLineBreak().enable(['heading', 'link', 'list', 'emphasis']);

  return markdown.render(value);
}
