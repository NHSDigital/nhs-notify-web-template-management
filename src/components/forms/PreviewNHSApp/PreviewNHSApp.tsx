'use client';

import { Radios } from 'nhsuk-react-components';
import { Preview } from '@molecules/Preview';
import { PreviewMessage } from '@organisms/PreviewMessage';
import { PreviewNHSAppProps } from './PreviewNHSApp.types';
import { renderMarkdown } from './server-actions';
import content from '@/src/content/content';

export function PreviewNHSApp({ templateName, message }: PreviewNHSAppProps) {
  const html = renderMarkdown(message);

  const {
    components: { previewNHSAppFormComponent },
  } = content;

  return (
    <div className='nhsuk-grid-row'>
      <div className='nhsuk-grid-column-two-thirds'>
        <PreviewMessage
          sectionHeading={previewNHSAppFormComponent.sectionHeader}
          templateName={templateName}
          details={previewNHSAppFormComponent.details}
          PreviewComponent={<Preview.NHSApp message={html} />}
          FormOptionsComponent={
            <Radios id='what-would-you-like-to-do-next' name='choice'>
              {previewNHSAppFormComponent.options.map((item, index) => (
                <Radios.Radio
                  data-testid={`preview-nhs-app-form__radios-${item.id}`}
                  key={`preview-nhs-app-form__radios-edit-${item.id}-${index}`}
                  value={item.id}
                >
                  {item.text}
                </Radios.Radio>
              ))}
            </Radios>
          }
        />
      </div>
    </div>
  );
}
