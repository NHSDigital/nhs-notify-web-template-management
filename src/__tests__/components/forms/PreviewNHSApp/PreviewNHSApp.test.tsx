import { render, screen } from '@testing-library/react';
import { mock } from 'jest-mock-extended';
import { PreviewNHSApp, PreviewNHSAppActions } from '@forms/PreviewNHSApp';
import { markdown } from '../fixtures';

describe('Preview nhs app form renders', () => {
  it('matches snapshot', () => {
    const container = render(
      <PreviewNHSApp
        pageActions={new PreviewNHSAppActions()}
        templateName='test-template-nhs app'
        message={markdown}
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('renders component correctly', () => {
    render(
      <PreviewNHSApp
        pageActions={new PreviewNHSAppActions()}
        templateName='test-template-nhs app'
        message='nhs app message body'
      />
    );

    expect(
      screen.getByTestId('preview-nhs-app-form__radios-edit')
    ).toHaveAttribute('value', 'edit');

    expect(
      screen.getByTestId('preview-nhs-app-form__radios-submit')
    ).toHaveAttribute('value', 'submit');
  });

  it('should should render message with markdown', () => {
    const message = 'nhs app message body';
    const pageActionsMock = mock<PreviewNHSAppActions>();

    pageActionsMock.renderMarkdown.mockReturnValue('Rendered via MD');

    render(
      <PreviewNHSApp
        pageActions={pageActionsMock}
        templateName='test-template-nhs app'
        message={message}
      />
    );

    expect(pageActionsMock.renderMarkdown).toHaveBeenCalledWith(message);

    expect(screen.getByTestId('preview__content-0')).toHaveTextContent(
      'Rendered via MD'
    );
  });
});
