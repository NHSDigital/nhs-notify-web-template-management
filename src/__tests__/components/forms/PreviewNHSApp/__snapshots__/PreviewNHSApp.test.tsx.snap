// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview nhs app form renders matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        data-testid="preview-message__heading"
      >
        <span
          class="nhsuk-caption-l"
          data-testid="preview-message__heading-caption"
        >
          NHS app message template
        </span>
        test-template-nhs app
      </h1>
      <details
        class="nhsuk-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="preview-message-details__heading"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Who your text message will be sent from
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="preview-message-details__text"
        >
          <p>
            Set your NHS App message sender name during onboarding.
          </p>
          <p>
            If you need to set up a different NHS App message sender name for other messages, contact our onboarding team.
          </p>
        </div>
      </details>
      <table
        class="nhsuk-table preview nhsuk-u-margin-bottom-4"
      >
        <tbody
          class="nhsuk-table__body"
        >
          <tr
            class="nhsuk-table__row"
            role="row"
          >
            <td
              class="nhsuk-table__cell"
              role="cell"
            >
              <div
                class="preview__heading"
                data-testid="preview__heading-0"
              >
                Message
              </div>
            </td>
            <td
              class="nhsuk-table__cell"
              role="cell"
            >
              <div
                class="preview__content"
                data-testid="preview__content-0"
              >
                <p>
                  message
                </p>
                

              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <form>
        <div
          class="nhsuk-form-group"
        >
          <fieldset
            class="nhsuk-fieldset"
          >
            <legend
              class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m"
              data-testid="preview-message-form__legend"
            >
              What would you like to do next with this template?
            </legend>
            <div
              class="nhsuk-form-group"
            >
              <div
                class="nhsuk-radios"
                id="what-would-you-like-to-do-next"
              >
                <div
                  class="nhsuk-radios__item"
                >
                  <input
                    aria-labelledby="what-would-you-like-to-do-next-1--label"
                    class="nhsuk-radios__input"
                    data-testid="preview-nhs-app-form__radios-edit"
                    id="what-would-you-like-to-do-next-1"
                    name="choice"
                    type="radio"
                    value="edit"
                  />
                  <label
                    class="nhsuk-label nhsuk-radios__label"
                    for="what-would-you-like-to-do-next-1"
                    id="what-would-you-like-to-do-next-1--label"
                  >
                    Edit
                  </label>
                </div>
                <div
                  class="nhsuk-radios__item"
                >
                  <input
                    aria-labelledby="what-would-you-like-to-do-next-2--label"
                    class="nhsuk-radios__input"
                    data-testid="preview-nhs-app-form__radios-submit"
                    id="what-would-you-like-to-do-next-2"
                    name="choice"
                    type="radio"
                    value="submit"
                  />
                  <label
                    class="nhsuk-label nhsuk-radios__label"
                    for="what-would-you-like-to-do-next-2"
                    id="what-would-you-like-to-do-next-2--label"
                  >
                    Submit
                  </label>
                </div>
              </div>
            </div>
          </fieldset>
        </div>
        <button
          aria-disabled="false"
          class="nhsuk-button"
          data-testid="preview-message-form__button"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
